fix: no-multiple-actions-in-effects not reporting mapTo operators

rafaelss95 opened the pull request:
Fixes #61.

in addition to fix the error reported in #61, I also changed the selectors to search for the exact names, instead of part of them, to avoid a possible problem with custom operators and also added some more tests to cover for all operators in the context.

rafaelss95 commented on src/utils/selectors/index.ts:
```diff
@@ -28,8 +28,13 @@ export const storeActionReducerMap = `${ngModuleDecorator} ObjectExpression Prop
-const effectsOperator = `ClassProperty > CallExpression[callee.name='createEffect'] CallExpression[callee.name=/switchMap|concatMap|mergeMap|flatMap|exhaustMap/]`
+const createEffectExpression = `ClassProperty > CallExpression[callee.name='createEffect']`
 
-export const effectsArrowReturn = `${effectsOperator} > ArrowFunctionExpression > ArrayExpression`
+const mapOperators = '(concat|exhaust|flat|merge|switch)Map'
+const mapToOperators = '(concat|merge|switch)MapTo'
+const mapOperatorsExpression = `${createEffectExpression} CallExpression[callee.name=/^${mapOperators}$/]`
+const mapToOperatorsExpression = `${createEffectExpression} CallExpression[callee.name=/^${mapToOperators}$/]`
 
-export const effectsReturn = `${effectsOperator} ReturnStatement`
+export const effectsImplicitReturn = `${mapOperatorsExpression} > ArrowFunctionExpression > ArrayExpression, ${mapToOperatorsExpression} ArrayExpression`
```
I'm not sure about all of these variable names. That's what came to mind this time, feel free to suggest :)

timdeschryver commented on tests/rules/no-multiple-actions-in-effects.test.ts:
```diff
@@ -34,6 +34,54 @@ ruleTester().run(ruleName, rule, {
+      code: `
+        export class Effects {
+          eleven$ = createEffect(() =>
+            this.actions$.pipe(switchMapTop([foo()]))
```
```suggestion
            this.actions$.pipe(switchMapTo([foo()]))
```

timdeschryver commented on tests/rules/no-multiple-actions-in-effects.test.ts:
```diff
@@ -34,6 +34,54 @@ ruleTester().run(ruleName, rule, {
+      code: `
+        export class Effects {
+          twelve$ = createEffect(() =>
+            this.actions$.pipe(aconcatMapTo([bar()]))
```
```suggestion
            this.actions$.pipe(concatMapTo([bar()]))
```

timdeschryver commented on src/utils/selectors/index.ts:
```diff
@@ -28,8 +28,13 @@ export const storeActionReducerMap = `${ngModuleDecorator} ObjectExpression Prop
+const createEffectExpression = `ClassProperty > CallExpression[callee.name='createEffect']`
 
-export const effectsArrowReturn = `${effectsOperator} > ArrowFunctionExpression > ArrayExpression`
+const mapOperators = '(concat|exhaust|flat|merge|switch)Map'
+const mapToOperators = '(concat|merge|switch)MapTo'
```
Nice!

timdeschryver reviewed (changes requested):
I spotted some typos, for the rest it's good - thanks!

timdeschryver commented:
Could we also add a test for an invalid `mapTo` ? 
I'm asking because I'm not 100% sure if we catch them with the used RegExp.

rafaelss95 commented:
Hey @timdeschryver, about the _typos_:

Well, actually I did this intentionally :stuck_out_tongue_closed_eyes:.

I've described in the PR description:

> I also changed the selectors to search for the exact names, instead of part of them, to avoid a possible problem with custom operators and also added some more tests to cover for all operators in the context.

Let me know if you got it or if you think it's unnecessary so I can remove it.

> Could we also add a test for an invalid `mapTo` ?
> I'm asking because I'm not 100% sure if we catch them with the used RegExp.

Hmm, should this rule report `map` and `mapTo`? Shouldn't we only report the operators that return observable? Let me know so I can change it appropriately.

timdeschryver commented:
Woops 😅
To be honest, I don't really use the `mapTo` operator in effects, but I suppose there will be usages of `mapTo` that return multiple actions. If we add the `mapTo` operators, perhas we should also add the "normal" `mapTo` operator?

rafaelss95 commented:
So, we can't return `mapTo` in a `effect`:

![Captura de Tela 2021-01-20 às 09 49 43](https://user-images.githubusercontent.com/11965907/105177060-e088f180-5b04-11eb-8780-c00265918326.png)
![Captura de Tela 2021-01-20 às 09 50 50](https://user-images.githubusercontent.com/11965907/105177128-00201a00-5b05-11eb-8f23-328fa5190dad.png)

timdeschryver commented:
Then let's get this merged! Thanks a lot

timdeschryver commented:
@all-contributors please add @rafaelss95  for code, tests
